### ISSUE TITLE
A socket timeout exception in Python 3 shows also a TypeError

### ISSUE THREAD

**Issue by xZise** [shazow/urllib3#682]
When a timout occurs it also shows the TypeError because you first try to call it with a `buffering` keyword argument, which does not exist in Python 3. I'm not sure if there is a good way in Python 3 to clear that TypeError exception. See also #537.

```
ERROR: Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 331, in _make_request
    httplib_response = conn.getresponse(buffering=True)
TypeError: getresponse() got an unexpected keyword argument 'buffering'
 
During handling of the above exception, another exception occurred:
 
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 333, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.4/http/client.py", line 1172, in getresponse
    response.begin()
  File "/usr/lib/python3.4/http/client.py", line 351, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.4/http/client.py", line 313, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib/python3.4/socket.py", line 371, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib/python3.4/ssl.py", line 745, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/lib/python3.4/ssl.py", line 617, in read
    v = self._sslobj.read(len, buffer)
socket.timeout: The read operation timed out
 
During handling of the above exception, another exception occurred:
 
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 362, in send
    timeout=timeout
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 559, in urlopen
    _pool=self, _stacktrace=stacktrace)
  File "/usr/lib/python3/dist-packages/urllib3/util/retry.py", line 245, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/usr/lib/python3/dist-packages/six.py", line 625, in reraise
    raise value
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 516, in urlopen
    body=body, headers=headers)
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 336, in _make_request
    self, url, "Read timed out. (read timeout=%s)" % read_timeout)
urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='commons.wikimedia.org', port=443): Read timed out. (read timeout=30)
 
During handling of the above exception, another exception occurred:
 
Traceback (most recent call last):
  File "/home/matanya/core/pywikibot/data/api.py", line 1911, in submit
    body=body, headers=headers)
  File "/home/matanya/core/pywikibot/tools/__init__.py", line 1227, in wrapper
    return obj(*__args, **__kw)
  File "/home/matanya/core/pywikibot/comms/http.py", line 240, in request
    r = fetch(baseuri, method, body, headers, **kwargs)
  File "/home/matanya/core/pywikibot/comms/http.py", line 359, in fetch
    error_handling_callback(request)
  File "/home/matanya/core/pywikibot/comms/http.py", line 276, in error_handling_callback
    raise request.data
  File "/home/matanya/core/pywikibot/comms/http.py", line 255, in _http_process
    auth=auth, timeout=timeout, verify=True)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 457, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 569, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 422, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='commons.wikimedia.org', port=443): Read timed out. (read timeout=30)
```

**Comment by Lukasa** [comment]
Yeah, this is a known issue that causes lots of bug reports on Requests. There is *no* way to do anything about this, though: the CPython developers believe this is a feature, not a bug.

**Comment by xZise** [comment]
Hmm that is unfortunate. Now in theory there is one (relatively) ugly workaround:

```
>>> def div(a, b):
...   try:
...     r = a / b
...   except TypeError:
...     r = None
...   if r is None:
...     r = a / int(b)
...   return r
... 
>>> div(42, 0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in div
ZeroDivisionError: division by zero
>>> div(42, '0')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 7, in div
ZeroDivisionError: division by zero
```

**Comment by jayvdb** [comment]
Any feedback on @xZise's workaround, which suggests moving the fallback code out of the exception block to avoid unnecessary exceptions being present in the reported exception?

**Comment by sigmavirus24** [comment]
@jayvdb I don't think there's an reasonable way to preserve the clarity of the code and to achieve that goal. These exception messages will be improved in later versions of Python 3, but that's not urllib3's problem.

**Comment by jayvdb** [comment]
This was fixed in #861
